// hibiki/dist/index.js
"use strict";
// index.ts - export all the things (noud02)
Object.defineProperty(exports, "__esModule", { value: true });
try {
    global.Promise = require("bluebird"); // tslint:disable-line:no-var-requires
}
catch (e) {
}
var Shard_1 = require("./client/Shard");
exports.Shard = Shard_1.Shard;
var WebSocketClient_1 = require("./client/WebSocketClient");
exports.WebSocketClient = WebSocketClient_1.WebSocketClient;
var Logger_1 = require("./client/Logger");
exports.Logger = Logger_1.Logger;
var PostgreSQL_1 = require("./db/PostgreSQL");
exports.PostgreSQL = PostgreSQL_1.PostgreSQL;
const ext = require("./ext/ExtDeco");
exports.ext = ext;
var Command_1 = require("./ext/Command");
exports.Command = Command_1.Command;
var CommandHandler_1 = require("./ext/CommandHandler");
exports.CommandHandler = CommandHandler_1.CommandHandler;
var Context_1 = require("./ext/Context");
exports.Context = Context_1.Context;
var ExtensionManager_1 = require("./ext/ExtensionManager");
exports.ExtensionManager = ExtensionManager_1.ExtensionManager;
var Ratelimiter_1 = require("./ext/Ratelimiter");
exports.Ratelimiter = Ratelimiter_1.Ratelimiter;
var LocaleManager_1 = require("./locale/LocaleManager");
exports.LocaleManager = LocaleManager_1.LocaleManager;
